pkg/koordlet/resourceexecutor/cgroup.go

```diff
@@ -0,0 +1,185 @@
+	if supported, msg := r.IsSupported(cgroupTaskDir); !supported {
+		return sysutil.ResourceUnsupportedErr(fmt.Sprintf("write cgroup %s failed, msg: %s", r.ResourceType(), msg))
+	}
+	if valid, msg := r.IsValid(value); !valid {
+		return fmt.Errorf("write cgroup %s failed, value[%v] not valid, msg: %s", r.ResourceType(), value, msg)
+	}
+	if exist, msg := sysutil.IsCgroupPathExist(cgroupTaskDir, r); !exist {
+		return sysutil.ResourceCgroupDirErr(fmt.Sprintf("write cgroup %s failed, msg: %s", r.ResourceType(), msg))
+	}
+
+	filePath := r.Path(cgroupTaskDir)
+	klog.V(5).Infof("write %s [%s]", filePath, value)
+
+	return os.WriteFile(filePath, []byte(value), 0644)
+}
+
+// CgroupFileReadInt reads the cgroup file and returns an int64 value.
+func cgroupFileReadInt(cgroupTaskDir string, r sysutil.Resource) (*int64, error) {
+	if supported, msg := r.IsSupported(cgroupTaskDir); !supported {
+		return nil, sysutil.ResourceUnsupportedErr(fmt.Sprintf("read cgroup %s failed, msg: %s", r.ResourceType(), msg))
+	}
+
+	dataStr, err := cgroupFileRead(cgroupTaskDir, r)
+	if err != nil {
+		return nil, err
+	}
+	if dataStr == "" {
+		return nil, fmt.Errorf(EmptyValueError)
+	}
+	if dataStr == CgroupMaxSymbolStr {
+		// compatible with cgroup valued "max"
+		data := int64(math.MaxInt64)
+		klog.V(6).Infof("read %s and got str value, considered as MaxInt64", r.Path(cgroupTaskDir))
+		return &data, nil
+	}
+	data, err := strconv.ParseInt(strings.TrimSpace(dataStr), 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	return &data, nil
+}
+
+// CgroupFileRead reads the cgroup file.
+func cgroupFileRead(cgroupTaskDir string, r sysutil.Resource) (string, error) {
+	if supported, msg := r.IsSupported(cgroupTaskDir); !supported {
+		return "", sysutil.ResourceUnsupportedErr(fmt.Sprintf("read cgroup %s failed, msg: %s", r.ResourceType(), msg))
+	}
+	if exist, msg := sysutil.IsCgroupPathExist(cgroupTaskDir, r); !exist {
+		return "", sysutil.ResourceCgroupDirErr(fmt.Sprintf("read cgroup %s failed, msg: %s", r.ResourceType(), msg))
+	}
+
+	filePath := r.Path(cgroupTaskDir)
+	klog.V(5).Infof("read %s", filePath)
+
+	data, err := os.ReadFile(filePath)
+	return strings.Trim(string(data), "\n"), err
+}
+
+func readCgroupAndParseInt64(parentDir string, r sysutil.Resource) (int64, error) {
+	s, err := cgroupFileRead(parentDir, r)
+	if err != nil && sysutil.IsCgroupDirErr(err) {
+		return -1, err
+	} else if err != nil {
+		return -1, fmt.Errorf("cannot read cgroup file, err: %v", err)
+	}
+
+	// "max" means unlimited
+	if strings.Trim(s, "\n ") == CgroupMaxSymbolStr {
+		return -1, nil
+	}
+	// content: `%lld`
+	v, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return -1, fmt.Errorf("cannot parse cgroup value %s, err: %v", s, err)
+	}
+	return v, nil
+}
+
+func readCgroupAndParseUint64(parentDir string, r sysutil.Resource) (uint64, error) {
+	s, err := cgroupFileRead(parentDir, r)
+	if err != nil {
```

Comment:
should not wrap the error since it can be ResourceUnsupportedErr or ResourceCgroupDirErr